File: training/3.2/sweet_butter/butter_dijkstra.py
"""
PROG: butter
LANG: PYTHON3

"""
import sys
import queue

with open('butter.in', 'r') as fin:
  lines = fin.readlines()

def dijkstra(edges, d, ind):
    q = queue.PriorityQueue()
    q.put([0, ind])
    visited = [False] * P
    while not q.empty():
        cur = q.get()
        #print("q get:", cur[0])
        if visited[cur[1]] == True:
            continue
        for edge in edges[cur[1]]:
            if d[ind][edge[1]] > d[ind][cur[1]] + edge[0]:
                d[ind][edge[1]] = d[ind][cur[1]] + edge[0]
                q.put([d[ind][edge[1]], edge[1]])
                #print("q put:", d[ind][edge[1]])
        visited[cur[1]] = True
    print("ind=", ind)
    for i in range(len(d[0])):
        print(d[ind][i])
    

N, P, C = map(int, lines[0].split())

pastures = []
for i in range(N):
    pastures.append(int(lines[i+1]) - 1)

edges = {}
for i in range(C):
    p1, p2, lenP = map(int, lines[N+1+i].split())
    p1n = p1 - 1
    p2n = p2 -1
    if p1n not in edges.keys():
        edges[p1n] = []
    edges[p1n].append([lenP, p2n])
    if p2n not in edges.keys():
        edges[p2n] = []
    edges[p2n].append([lenP, p1n])
    
d = [[sys.maxsize for i in range(P)] for i in range(P)]

for i in range(P):
    d[i][i] = 0

for i in pastures:
    dijkstra(edges, d, i)

result = sys.maxsize
for i in range(P):
    totalLen = 0
    for pasture in pastures:
        totalLen += d[pasture][i]
        if i == 0:
            print("pasture=", pasture, "d=", d[pasture][i])
    if result > totalLen:
        result = totalLen
        print("i=", i, ", result=", result)

with open('butter.out', 'w') as fout:
    fout.write(str(result) + '\n')

File: training/3.2/sweet_butter/butter_dijkstra.cpp
/*
PROG: butter
LANG: C++
*/

#include <iostream>
#include <fstream>
#include <vector>
#include <queue>

using namespace std;

#define MYMAX (1 << 30)
#define PASTURES 800
#define COWS 500

struct Edge {
    int pasture;
    int distance;
    Edge(int p, int d) : pasture(p), distance(d) {}
    bool operator<(const Edge& other) const
    {
        return distance > other.distance; // opposite > because top is biggest by default
    } 
};

int d[PASTURES][PASTURES];
int pastures[COWS];
vector<Edge> edges[PASTURES];

void dijkstra(int u, int P, const vector<Edge> edges[PASTURES], int d[PASTURES][PASTURES])
{
    priority_queue<Edge> q;
    q.push(Edge(u, 0));
    bool visited[PASTURES];
    //cout << "u=" << u << endl;
    for (int i = 0; i < P; i++)
        visited[i] = false;
    while(!q.empty())
    {
        Edge cur = q.top();
        //cout << "cur=" << cur << endl;
        q.pop();
        if (visited[cur.pasture] == true)
            continue;
        visited[cur.pasture] = true;
        for (int i = 0; i < edges[cur.pasture].size(); ++i)
        {
            int v = edges[cur.pasture][i].pasture;
            int distance = edges[cur.pasture][i].distance;
            //if (cur == 3 && v == 6)
                //cout << "d[0][6]=" << d[0][6] << ", d[0][3]=" << d[0][3] << ", w[3][6]=" << w[3][6] << ", visited[6]=" << visited[6] << endl;
            if (d[u][v] > d[u][cur.pasture] + distance)
            {
                d[u][v] = d[u][cur.pasture] + distance;

                if (visited[v] == false)
                {
                    q.push(Edge(v, d[u][v]));
                    //cout << "pushing=" << v << endl;
                }
            }
        }
    }
}

int main()
{
    ifstream fin("butter.in");
    ofstream fout("butter.out");

    int N, P, C;
    fin >> N >> P >> C;

    for (int i = 0; i < N; ++i)
    {
        fin >> pastures[i];
        pastures[i] -= 1;
    }

    for (int i = 0; i < P; ++i)
        for (int j = 0; j < P; ++j)
        {
            d[i][j] = MYMAX;
        }

    for (int i = 0; i < C; ++i)
    {
        int start, end, size;
        fin >> start >> end >> size;
        
        edges[start-1].push_back(Edge(end-1, size));
        edges[end-1].push_back(Edge(start-1, size));
    }

    for (int i = 0; i < P; ++i)
    {
        d[i][i] = 0;
    }

    for (int i=0; i < P; ++i)
    {
        dijkstra(i, P, edges, d);
    }
    int result = MYMAX;
    for (int i = 0; i < P; ++i)
    {
        int totalLen = 0;
        for (int j = 0; j < N; ++j)
            totalLen += d[pastures[j]][i];
        if (result > totalLen)
        {
            result = totalLen;
            //cout << "i=" << i << endl;
        }
    }

    fout << result << endl;
}